### PR TITLE
Warn for queue permissions for everyone and anonymous only if access rights are set to allow

### LINKED ISSUE #4610: MSMQ: The TryGetPermissions function in MessageQueueExtensions does not take the ACEType into account
Currently the TryGetPermissions method in the MessageQueueExtensions class determines what message queue access rights are present for a given user Sid. However, the method does not return if the user has Allow access or Deny access.

### PR DESCRIPTION
**Pull request by indualagarsamy**
Fixes: https://github.com/Particular/NServiceBus/issues/4526
Also fixes #4610

### CODE REVIEW

File: src/NServiceBus.Core.Tests/Msmq/QueuePermissionsTests.cs [hunk 1]
```diff
@@ -13,7 +13,7 @@ public class QueuePermissionsTests
     {
         StringBuilder logOutput;
         string testQueueName = "NServiceBus.Core.Tests.QueuePermissionsTests";
-
+    
         [OneTimeSetUp]
```
Comment (andreasohlund) [inline src/NServiceBus.Core.Tests/Msmq/QueuePermissionsTests.cs]: Making this [Teardown] should fix the issue `OneTimeSetUp` seems to be a misstake?

Comment (indualagarsamy) [inline src/NServiceBus.Core.Tests/Msmq/QueuePermissionsTests.cs]: @kbaley intended for this to be a OneTimeSetup for the logger.  However, the problem was that the in the the log string wasn't being cleared on teardown.

**Comment by indualagarsamy** [comment]
@gbiellem @andreasohlund - this is ready for review.

File: src/NServiceBus.Core.Tests/Msmq/MsmqExtensionsTests.cs [hunk 2]
```diff
@@ -33,18 +32,22 @@ public void Teardown()
         }
 
         [Test]
```
Comment (andreasohlund) [inline src/NServiceBus.Core.Tests/Msmq/MsmqExtensionsTests.cs]: You can remove this

File: src/NServiceBus.Core.Tests/Msmq/MsmqExtensionsTests.cs [hunk 3]
```diff
@@ -33,18 +32,22 @@ public void Teardown()
...
+        public void GetPermissions_returns_queue_access_rights(AccessControlEntryType providedAccessType)
         {
-            queue.SetPermissions(LocalEveryoneGroupName, MessageQueueAccessRights.WriteMessage | MessageQueueAccessRights.ReceiveMessage, AccessControlEntryType.Allow);
+            queue.SetPermissions(LocalEveryoneGroupName, MessageQueueAccessRights.WriteMessage | MessageQueueAccessRights.ReceiveMessage, providedAccessType);
             MessageQueueAccessRights? rights;
-            if (!queue.TryGetPermissions(LocalEveryoneGroupName, out rights))
+            AccessControlEntryType? accessType;
+            if (!queue.TryGetPermissions(LocalEveryoneGroupName, out rights, out accessType))
             {
                 Assert.Fail("Unable to read permissions off a queue");
```
Comment (andreasohlund) [inline src/NServiceBus.Core.Tests/Msmq/MsmqExtensionsTests.cs]: Nitpick: print the qureu name in the message?

File: src/NServiceBus.Core.Tests/Msmq/MsmqQueueCreatorTests.cs [hunk 4]
```diff
@@ -63,17 +63,23 @@ public void Should_setup_permissions()
...
-
-            Assert.True(createdQueue.TryGetPermissions(NetworkServiceAccountName, out accountAccessRights));
-            Assert.True(accountAccessRights.HasValue, "User should have access rights");
-            Assert.AreEqual(MessageQueueAccessRights.WriteMessage | MessageQueueAccessRights.ReceiveMessage, accountAccessRights, "User should have write/read access");
-
+            AccessControlEntryType? accessControlEntryType;
+
+            Assert.True(createdQueue.TryGetPermissions(NetworkServiceAccountName, out accountAccessRights, out accessControlEntryType));
+            Assert.True(accountAccessRights.HasValue);
+            Assert.True(accessControlEntryType == AccessControlEntryType.Allow,"User should have access");
```
Comment (andreasohlund) [inline src/NServiceBus.Core.Tests/Msmq/MsmqQueueCreatorTests.cs]: Some whitespace missing, do a R# "Format My Code Using "Particular" conventions"

File: src/NServiceBus.Core.Tests/Msmq/MsmqQueueCreatorTests.cs [hunk 5]
```diff
@@ -63,17 +63,23 @@ public void Should_setup_permissions()
...
 
-            Assert.True(createdQueue.TryGetPermissions(LocalAdministratorsGroupName, out localAdminAccessRights));
-            Assert.True(localAdminAccessRights.HasValue, "User should have access rights");
-            Assert.AreEqual(MessageQueueAccessRights.FullControl, localAdminAccessRights, "LocalAdmins should have full control");
-        }
+
+            Assert.True(createdQueue.TryGetPermissions(LocalAdministratorsGroupName, out localAdminAccessRights, out accessControlEntryTypeForLocalAdmin));
+            Assert.True(localAdminAccessRights.HasValue);
+            Assert.True(localAdminAccessRights?.HasFlag(MessageQueueAccessRights.FullControl), "LocalAdmins should have full control");
+            Assert.IsTrue(accessControlEntryTypeForLocalAdmin == AccessControlEntryType.Allow, "User should have access");
```
Comment (andreasohlund) [inline src/NServiceBus.Core.Tests/Msmq/MsmqQueueCreatorTests.cs]: User == Local Admin?

File: src/NServiceBus.Core.Tests/Msmq/MsmqQueueCreatorTests.cs [hunk 6]
```diff
@@ -125,17 +131,20 @@ public void Should_give_everyone_and_anonymous_access_rights_when_creating_queue
...
-
+                Assert.True(everyoneAccessRights?.HasFlag(MessageQueueAccessRights.GenericWrite), "Msmq should give 'everyone' write access by default");
+                Assert.True(accessControlEntryTypeForEveryone == AccessControlEntryType.Allow);
+                
                 MessageQueueAccessRights? anonymousAccessRights;
+                AccessControlEntryType? accessControlEntryTypeForAnonymous;
 
-                Assert.True(queue.TryGetPermissions(LocalAnonymousLogonName, out anonymousAccessRights));
+                Assert.True(queue.TryGetPermissions(LocalAnonymousLogonName, out anonymousAccessRights, out accessControlEntryTypeForAnonymous));
                 Assert.True(anonymousAccessRights.HasValue, "User should have access rights");
```
Comment (andreasohlund) [inline src/NServiceBus.Core.Tests/Msmq/MsmqQueueCreatorTests.cs]: User => anonymous user?

File: src/NServiceBus.Core.Tests/Msmq/MsmqQueueCreatorTests.cs [hunk 7]
```diff
@@ -125,17 +131,20 @@ public void Should_give_everyone_and_anonymous_access_rights_when_creating_queue
             using (var queue = MessageQueue.Create(path))
             {
                 MessageQueueAccessRights? everyoneAccessRights;
+                AccessControlEntryType? accessControlEntryTypeForEveryone;
 
-                Assert.True(queue.TryGetPermissions(LocalEveryoneGroupName, out everyoneAccessRights));
+                Assert.True(queue.TryGetPermissions(LocalEveryoneGroupName, out everyoneAccessRights, out accessControlEntryTypeForEveryone));
                 Assert.True(everyoneAccessRights.HasValue, "User should have access rights");
```
Comment (andreasohlund) [inline src/NServiceBus.Core.Tests/Msmq/MsmqQueueCreatorTests.cs]: User => Everyone

File: src/NServiceBus.Core.Tests/Msmq/QueuePermissionsTests.cs [hunk 8]
```diff
@@ -22,21 +22,39 @@ public void TestFixtureSetup()
             logOutput = new StringBuilder();
             var stringWriter = new StringWriter(logOutput);
             loggerFactory.WriteTo(stringWriter);
+
```
Comment (andreasohlund) [inline src/NServiceBus.Core.Tests/Msmq/QueuePermissionsTests.cs]: Nitpick: extra line ( a R# format should do the trick)

File: src/NServiceBus.Core.Tests/Msmq/QueuePermissionsTests.cs [hunk 9]
```diff
@@ -22,21 +22,39 @@ public void TestFixtureSetup()
...
+ 
+        [Test]
+        public void Should_not_warn_if_queue_has_public_access_set_to_deny()
+        {
+            // Set up a queue with read/write access for everyone/anonymous explicitly set to DENY. 
+            var everyoneGroupName = new SecurityIdentifier(WellKnownSidType.WorldSid, null).Translate(typeof(NTAccount)).ToString();
+            var anonymousGroupName = new SecurityIdentifier(WellKnownSidType.AnonymousSid, null).Translate(typeof(NTAccount)).ToString();
+            using (var queue = MessageQueue.Create(@".\private$\" + testQueueName, false))
+            {
+                queue.SetPermissions(everyoneGroupName, MessageQueueAccessRights.GenericRead, AccessControlEntryType.Deny);
```
Comment (andreasohlund) [inline src/NServiceBus.Core.Tests/Msmq/QueuePermissionsTests.cs]: There are more queue access rights do we need to check any of those?

https://msdn.microsoft.com/en-us/library/system.messaging.messagequeueaccessrights(v=vs.110).aspx

File: src/NServiceBus.Core.Tests/Msmq/QueuePermissionsTests.cs [hunk 10]
```diff
@@ -66,18 +84,22 @@ public void Should_log_if_queue_exists()
...
 
-        [Test]
-        public void Should_warn_if_queue_has_public_access()
+        [TestCase(MessageQueueAccessRights.FullControl, WellKnownSidType.WorldSid)]
+        [TestCase(MessageQueueAccessRights.GenericRead, WellKnownSidType.WorldSid)]
+        [TestCase(MessageQueueAccessRights.GenericWrite, WellKnownSidType.WorldSid)]
+        [TestCase(MessageQueueAccessRights.FullControl, WellKnownSidType.AnonymousSid)]
+        [TestCase(MessageQueueAccessRights.GenericRead, WellKnownSidType.AnonymousSid)]
+        [TestCase(MessageQueueAccessRights.GenericWrite, WellKnownSidType.AnonymousSid)]
+        public void Should_warn_if_queue_has_public_access(MessageQueueAccessRights rights, WellKnownSidType sidType)
```
Comment (andreasohlund) [inline src/NServiceBus.Core.Tests/Msmq/QueuePermissionsTests.cs]: Same here there are more options. should we check those? (Like `DeleteQueue` https://msdn.microsoft.com/en-us/library/system.messaging.messagequeueaccessrights(v=vs.110).aspx )

File: src/NServiceBus.Core/Transports/Msmq/QueuePermissions.cs [hunk 11]
```diff
@@ -43,34 +42,30 @@ public static void CheckQueue(string address)
...
 
-                if (Debugger.IsAttached)
-                {
-                    Logger.Info(logMessage);
-                }
-                else
-                {
-                    Logger.Warn(logMessage);
-                }
+            if (accessRights?.HasFlag(MessageQueueAccessRights.GenericRead) == true ||
```
Comment (andreasohlund) [inline src/NServiceBus.Core/Transports/Msmq/QueuePermissions.cs]: I know we talked about making this generic and WARN if there are any `Allow` AccessControlEntry for the public groups. Was that considered?

That would make the code much cleaner and would save us from checking other access rights?

Eg. a security expert would probably not like that we skip the `DeleteQueue` AccessControlEntryType?

File: src/NServiceBus.Core/Utils/MessageQueueExtensions.cs [hunk 12]
```diff
@@ -75,7 +75,15 @@ enum ACL_INFORMATION_CLASS
...
-        public static bool TryGetPermissions(this MessageQueue queue, string user, out MessageQueueAccessRights? rights)
+        /// <summary>
+        /// Using interop, determines the access control rights for the specified user.
+        /// </summary>
+        /// <param name="queue">MSMQ</param>
+        /// <param name="user">user account name</param>
+        /// <param name="rights">determines the access rights defined for this queue</param>
+        /// <param name="accessType">determines if the access is granted or deny.</param>
+        /// <returns>true if successful, false otherwise</returns>
+        public static bool TryGetPermissions(this MessageQueue queue, string user, out MessageQueueAccessRights? rights, out AccessControlEntryType? accessType)
```
Comment (andreasohlund) [inline src/NServiceBus.Core/Utils/MessageQueueExtensions.cs]: We usually don't put API Docs on non public methods?

File: src/NServiceBus.Core.Tests/Msmq/MsmqExtensionsTests.cs (hunk 3 above)
Comment (indualagarsamy) [inline src/NServiceBus.Core.Tests/Msmq/MsmqExtensionsTests.cs]: @andreasohlund - I can do that, just FYI, I did not make changes to that code. I do agree having the queue name will make it more easier as well.

File: src/NServiceBus.Core.Tests/Msmq/MsmqQueueCreatorTests.cs (hunk 6 above)
Comment (indualagarsamy) [inline src/NServiceBus.Core.Tests/Msmq/MsmqQueueCreatorTests.cs]: I will fix the message displayed as well to include the specified username.

File: src/NServiceBus.Core.Tests/Msmq/MsmqQueueCreatorTests.cs (hunk 7 above)
Comment (indualagarsamy) [inline src/NServiceBus.Core.Tests/Msmq/MsmqQueueCreatorTests.cs]: I will fix the message displayed as well to include the specified username.

File: src/NServiceBus.Core.Tests/Msmq/MsmqQueueCreatorTests.cs (hunk 5 above)
Comment (indualagarsamy) [inline src/NServiceBus.Core.Tests/Msmq/MsmqQueueCreatorTests.cs]: I will fix the message displayed as well to include the specified username.

File: src/NServiceBus.Core/Transports/Msmq/QueuePermissions.cs (hunk 11 above)
Comment (indualagarsamy) [inline src/NServiceBus.Core/Transports/Msmq/QueuePermissions.cs]: @andreasohlund - Yeah, will do. Now that I have made the necessary changes to expose the AclType it should be easy. I meant to make that change :(

**Comment by indualagarsamy** [comment]
@andreasohlund - I have addressed the changes you brought up. Once this PR is merged, I have a few other changes I'd like to make. Currently some of the MSMQ unit tests don't clean up after themselves, they leave queues behind etc. But I'd like to address those in a separate PR.

**Comment by andreasohlund** [comment]
@bording @dbelcham can you review and merge if good

File: src/NServiceBus.Msmq.AcceptanceTests/When_starting_up_the_endpoint.cs [hunk 13]
```diff
@@ -34,17 +33,15 @@ class Context : ScenarioContext
...
                 if (initialized)
                 {
                     return;
                 }
                 initialized = true;
-                EndpointSetup<DefaultServer>(c =>
-                {
-                    c.UseTransport<MsmqTransport>();
-                });
+                EndpointSetup<DefaultServer>(c => { c.UseTransport<MsmqTransport>(); });
```
Comment (timbussmann) [inline src/NServiceBus.Msmq.AcceptanceTests/When_starting_up_the_endpoint.cs]: transport configuration is not necessary

**Comment by andreasohlund** [comment]
@timbussmann made the change you mentioned, merging

**Pull request merged**